environ/funcs.go

```diff
@@ -121,3 +126,72 @@ func sub(x int, vals ...int) int {
 	}
 	return x
 }
+
+func plugin(name, function string, ctx interface{}) (interface{}, error) {
+	return callPlugin(simpleRunner, name, function, ctx)
+}
+
+func callPlugin(runner cmdRunner, name, function string, ctx interface{}) (interface{}, error) {
+	d := make(map[string]interface{})
+	d["data"] = ctx
+	b, err := json.Marshal(d)
+	if err != nil {
+		return nil, err
+	}
+	o, err := execJSON(runner, name, b, function)
+	if err != nil {
+		return nil, err
+	}
+	return o["data"], nil
+}
+
+type cmdRunner func(name string, args ...string) *exec.Cmd
+
+func simpleRunner(name string, args ...string) *exec.Cmd {
+	return exec.Command(name, args...)
+}
+
+// execJSON executes the plugin name with arguments args. It creates a io.Pipe
+// to stdin of the command and writes the data into the pipe in a separate
+// goroutine.
+//
+// The output is decoded as json data into a map[string]interface{}
+func execJSON(runner cmdRunner, name string, data []byte, args ...string) (map[string]interface{}, error) {
+	cmd := runner(name, args...)
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		defer stdin.Close()
+		stdin.Write(data)
+	}()
+	if err != nil {
+		return nil, err
+	}
+	v, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, errors.New(err.Error() + string(v))
```

Comment:
this should probably be errors.Wrapf(err, "error running plugin %q: " string(v))